### ISSUE TITLE
push/pull: support an aggregate path as a target that has multiple outputs inside

### ISSUE THREAD

**Issue by skshetry** [iterative/dvc#10286]
Let's say we have the following hierarchy, where `data/bar` and `data/foo` are individual outputs.
When you do `dvc push data`, dvc does not know to search for outputs inside the path.

```console
.
└── data
   ├── bar
   └── foo
```

At the moment, `target` only supports filtering subpath of an output or an exact path to the output.

So, it would work for `data/foo` or `data/bar`. And, if the `data` was being tracked as single output, `data/bar` and `data/foo` will work to filter the `data` output.

---

Regarding implementation, we have internal utilities to collect all outputs of a path using `find_outs_by_path(recursive=True)`. But since it can return multiple stages, there may be assumptions that it can return only a single output/stage which have to be fixed. And, more importantly, test that it does not affect other commands, since this `collect_granular` is used in a lot of places.

https://github.com/iterative/dvc/blob/44b78b81b99bac06bf4549487656a5d20005f5b6/dvc/repo/stage.py#L396

Related: #10203.

**Comment by dberenbaum** [comment]
@skshetry Is it only about push/pull? Does it impact checkout or other commands?

**Comment by skshetry** [comment]
Yes, it likely affects all of the commands, as most of them `collect_granular`.
